### ISSUE TITLE
The 'sqlalchemy<1.5,>=1.3.0' distribution was not found and is required by buildbot

### ISSUE THREAD

**Issue by ryandesign** [buildbot/buildbot#7053]
Attempting to run buildbot 3.9.0 fails with:

```
% buildbot --version
Traceback (most recent call last):
  File "/opt/local/Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages/pkg_resources/__init__.py", line 631, in _build_master
    ws.require(__requires__)
  File "/opt/local/Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages/pkg_resources/__init__.py", line 968, in require
    needed = self.resolve(parse_requirements(requirements))
  File "/opt/local/Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages/pkg_resources/__init__.py", line 829, in resolve
    dist = self._resolve_dist(
  File "/opt/local/Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages/pkg_resources/__init__.py", line 875, in _resolve_dist
    raise VersionConflict(dist, req).with_context(dependent_req)
pkg_resources.ContextualVersionConflict: (SQLAlchemy 2.0.19 (/opt/local/Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages), Requirement.parse('sqlalchemy<1.5,>=1.3.0'), {'buildbot'})

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/opt/local/bin/buildbot", line 33, in <module>
    sys.exit(load_entry_point('buildbot==3.9.0', 'console_scripts', 'buildbot')())
  File "/opt/local/bin/buildbot", line 25, in importlib_load_entry_point
    return next(matches).load()
  File "/opt/local/Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/importlib/metadata/__init__.py", line 171, in load
    module = import_module(match.group('module'))
  File "/opt/local/Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/importlib/__init__.py", line 126, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
  File "<frozen importlib._bootstrap>", line 1050, in _gcd_import
  File "<frozen importlib._bootstrap>", line 1027, in _find_and_load
  File "<frozen importlib._bootstrap>", line 1006, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 688, in _load_unlocked
  File "<frozen importlib._bootstrap_external>", line 883, in exec_module
  File "<frozen importlib._bootstrap>", line 241, in _call_with_frames_removed
  File "/opt/local/Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages/buildbot/scripts/runner.py", line 32, in <module>
    from buildbot.scripts import base
  File "/opt/local/Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages/buildbot/scripts/base.py", line 28, in <module>
    from buildbot.config.master import FileLoader
  File "/opt/local/Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages/buildbot/config/master.py", line 43, in <module>
    from buildbot.www import auth
  File "/opt/local/Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages/buildbot/www/auth.py", line 17, in <module>
    import pkg_resources
  File "/opt/local/Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages/pkg_resources/__init__.py", line 3328, in <module>
    def _initialize_master_working_set():
  File "/opt/local/Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages/pkg_resources/__init__.py", line 3302, in _call_aside
    f(*args, **kwargs)
  File "/opt/local/Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages/pkg_resources/__init__.py", line 3340, in _initialize_master_working_set
    working_set = WorkingSet._build_master()
  File "/opt/local/Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages/pkg_resources/__init__.py", line 633, in _build_master
    return cls._build_from_requirements(__requires__)
  File "/opt/local/Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages/pkg_resources/__init__.py", line 646, in _build_from_requirements
    dists = ws.resolve(reqs, Environment())
  File "/opt/local/Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages/pkg_resources/__init__.py", line 829, in resolve
    dist = self._resolve_dist(
  File "/opt/local/Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages/pkg_resources/__init__.py", line 870, in _resolve_dist
    raise DistributionNotFound(req, requirers)
pkg_resources.DistributionNotFound: The 'sqlalchemy<1.5,>=1.3.0' distribution was not found and is required by buildbot
```

This did not happen with buildbot 3.8.0, which said:

```
% buildbot --version
Buildbot version: 3.8.0
Twisted version: 22.10.0
```

I have sqlalchemy 2.0.19 installed.

**Comment by p12tic** [comment]
Sqlalchemy 2.0.19 is not supported.

**Comment by ryandesign** [comment]
Well, it is the latest version of sqlalchemy and the one available in MacPorts. What will it take to get it supported?

Why didn't 3.8.0 let me know it wasn't supported?

Why doesn't 3.9.0 say it isn't supported at build time? Why does it wait until runtime?

Why do new versions of sqlalchemy keep breaking buildbot? Can buildbot switch to something less prone to breaking?

**Comment by pmisik** [comment]
I guess Buildbot stayed on sqlalchemy 1.x due to Buildbt master supports Python 3.5|3.6|3.7... (see https://buildbot.net/#/)
Sqlalchemy 2.x is supported just since Python 3.7.

With the release of Buildbot 4.x it could be revised to drop support for unsupported versions 3.5, 3.6, 3.7, (3.8) as many packages already dropped support for unsupported Python 3 versions.

>What will it take to get it supported?

It looks like there API changes an there is migration guide 
https://docs.sqlalchemy.org/en/20/changelog/migration_20.html
Feel free and contribute to Buildbot project with PR that will adapt its codebase to be compatible with sqlalchemy 2.x

It looks like 1.4.x is still supported including building wheel package for macOS see https://github.com/sqlalchemy/sqlalchemy/releases/tag/rel_1_4_49

**Comment by p12tic** [comment]
> Why doesn't 3.9.0 say it isn't supported at build time? Why does it wait until runtime?

Was MacPorts used to install the packages? If yes, this would explain why error is not thrown at build time, because dependencies are handled by macports. If e.g. pip was used, it would install correct version of SQLAlchemy.

**Comment by ryandesign** [comment]
Yes, MacPorts was used to install buildbot and its dependencies. I expect the buildbot build process to notice if there is a problem and exit with an error, rather than allow the installation of software that doesn't work.

If sqlalchemy 2 will not be supported until buildbot 4, then I guess buildbot in MacPorts will be unusable until buildbot 4 is released.

**Comment by ryandesign** [comment]
Dependency information is of course handled by MacPorts to the extent that MacPorts ports indicate which other ports they depend upon, but MacPorts does not support the practice that Python software often uses of requiring specific versions of other software. MacPorts can only offer you whatever the current version of a port is. Therefore it is advisable to support the latest several versions of your dependencies.

MacPorts of course invokes setup.py to install Python software—or, it used to, until [PEP517](https://peps.python.org/pep-0517/) became the preferred method to install Python software, and now MacPorts normally uses that instead. MacPorts deliberately disables any attempts by the build system to install its own dependencies since MacPorts has already handled that. So buildbot can of course do whatever it wants during the installation process, including verifying that installed dependencies are of the correct version.

**Comment by p12tic** [comment]
Fixed in #7681.

cc @ryandesign

**Comment by p12tic** [comment]
This is not yet part of any release.
